weave/trace_server/llm_completion.py

```diff
@@ -124,3 +160,108 @@ def get_azure_credentials(model_name: str) -> tuple[str, str]:
         )
 
     return azure_api_base, azure_api_version
+
+
+def get_custom_provider_info(
+    project_id: str,
+    model_name: str,
+    obj_read_func: Callable,
+) -> tuple[str, str, dict[str, str], str, str]:
+    """
+    Extract provider information from a custom provider model.
+    Args:
+        project_id: The project ID
+        model_name: The model name (format: <provider_id>/<provider_model_id>)
+        obj_read_func: Function to read objects from the database
+        secret_fetcher: Secret fetcher to get API keys
+    Returns:
+        Tuple containing:
+        - base_url: The base URL for the provider
+        - api_key: The API key for the provider
+        - extra_headers: Extra headers to send with the request
+        - return_type: The return type for the provider
+        - actual_model_name: The actual model name to use for the API call
+    """
+    secret_fetcher = _secret_fetcher_context.get()
+    if not secret_fetcher:
+        raise InvalidRequest(
+            f"No secret fetcher found, cannot fetch API key for model {model_name}"
+        )
+
+    # Parse the model name to extract provider_id and provider_model_id
+    # Format: <provider_id>/<provider_model_id>
+    parts = model_name.split("/")
+    if len(parts) < 2:
+        raise InvalidRequest(f"Invalid custom provider model format: {model_name}")
+
+    provider_id = parts[0]
+    provider_model_id = parts[1]
+
+    # Default values
+    base_url = None
+    secret_name = None
+    extra_headers = {}
+    return_type = "openai"
+    actual_model_name = model_name
+
+    try:
+        # Fetch the provider object
+        provider_obj_req = tsi.ObjReadReq(
+            project_id=project_id,
+            object_id=provider_id,
+            digest="latest",
+            metadata_only=False,
+        )
+        provider_obj_res = obj_read_func(provider_obj_req)
+        provider_obj = provider_obj_res.obj
+
+        if provider_obj.base_object_class != "Provider":
+            raise InvalidRequest(
+                f"Object {provider_id} is not a Provider, it is a {provider_obj.base_object_class}"
+            )
+
+        # Extract provider information
```

Comment:
you could convert the payload into the `Provider` class (and same for the model below) which would reduce the use of raw dictionary access